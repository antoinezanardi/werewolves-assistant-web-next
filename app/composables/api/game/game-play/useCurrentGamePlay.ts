import type { ComputedRef, Ref } from "vue";
import type { GamePlayCause } from "~/composables/api/game/types/game-play/game-play.types";

import type { Game } from "~/composables/api/game/types/game.class";
import type { PlayerInteractionType } from "~/composables/api/game/types/players/player-interaction/player-interaction.types";
import type { Player } from "~/composables/api/game/types/players/player.class";

type UseCurrentGamePlay = {
  mustCurrentGamePlayBeSkipped: ComputedRef<boolean>;
  priorityCauseInCurrentGamePlay: ComputedRef<GamePlayCause | undefined>;
  getEligibleTargetsWithInteractionInCurrentGamePlay: (interaction: PlayerInteractionType) => Player[];
};

function useCurrentGamePlay(game: Ref<Game>): UseCurrentGamePlay {
  const mustCurrentGamePlayBeSkipped = computed<boolean>(() => {
    const isWolfHoundSideRandomlyChosen = game.value.options.roles.wolfHound.isSideRandomlyChosen;
    const stealRoleEligibleTargets = getEligibleTargetsWithInteractionInCurrentGamePlay("steal-role");
    const currentGameAction = game.value.currentPlay?.action;

    return currentGameAction === "bury-dead-bodies" && !stealRoleEligibleTargets.length ||
      currentGameAction === "choose-side" && isWolfHoundSideRandomlyChosen;
  });

  const priorityCauseInCurrentGamePlay = computed<GamePlayCause | undefined>(() => {
    const gamePlayCausesSortedByPriority: GamePlayCause[] = [
      "previous-votes-were-in-ties",
      "angel-presence",
      "stuttering-judge-request",
    ];

    return gamePlayCausesSortedByPriority.find(cause => game.value.currentPlay?.causes?.includes(cause));
  });

  function getEligibleTargetsWithInteractionInCurrentGamePlay(interactionType: PlayerInteractionType): Player[] {
    if (game.value.currentPlay?.source.interactions === undefined) {
      return [];
    }
    const interaction = game.value.currentPlay.source.interactions.find(({ type }) => type === interactionType);
    if (interaction === undefined) {
      return [];
    }
    return interaction.eligibleTargets;
  }
  return {
    mustCurrentGamePlayBeSkipped,
    priorityCauseInCurrentGamePlay,
    getEligibleTargetsWithInteractionInCurrentGamePlay,
  };
}

export { useCurrentGamePlay };