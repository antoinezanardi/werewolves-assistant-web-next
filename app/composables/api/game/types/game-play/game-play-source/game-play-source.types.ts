import type { TupleToUnion } from "type-fest";

import type { GAME_PLAY_SOURCE_NAMES } from "~/composables/api/game/constants/game-play/game-play-source/game-play-source.constants";

type GamePlaySourceName = TupleToUnion<typeof GAME_PLAY_SOURCE_NAMES>;

export type { GamePlaySourceName };