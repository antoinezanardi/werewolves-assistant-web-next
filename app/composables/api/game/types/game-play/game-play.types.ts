import type { TupleToUnion } from "type-fest";

import type { GAME_PLAY_TYPES, GAME_PLAY_CAUSES, GAME_PLAY_OCCURRENCES, WITCH_POTIONS, GAME_PLAY_ACTIONS } from "~/composables/api/game/constants/game-play/game-play.constants";

type GamePlayType = TupleToUnion<typeof GAME_PLAY_TYPES>;

type GamePlayAction = TupleToUnion<typeof GAME_PLAY_ACTIONS>;

type GamePlayCause = TupleToUnion<typeof GAME_PLAY_CAUSES>;

type GamePlayOccurrence = TupleToUnion<typeof GAME_PLAY_OCCURRENCES>;

type WitchPotion = TupleToUnion<typeof WITCH_POTIONS>;

export type {
  GamePlayType,
  GamePlayAction,
  GamePlayCause,
  GamePlayOccurrence,
  WitchPotion,
};