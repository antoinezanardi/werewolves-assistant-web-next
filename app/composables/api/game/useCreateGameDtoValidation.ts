import { storeToRefs } from "pinia";
import { group as groupByField } from "radash";
import type { ComputedRef, Ref } from "vue";

import { MIN_PLAYERS_IN_GAME, MIN_PLAYERS_IN_GROUP } from "~/composables/api/game/constants/game.constants";
import type { CreateGamePlayerWithGroupDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { Role } from "~/composables/api/role/types/role.class";
import { useRolesStore } from "~/stores/role/useRolesStore";

type RoleWithMinInGame = Role & { minInGame: number };

type UseCreateGameDtoValidation = {
  isMinimumPlayersReached: ComputedRef<boolean>;
  areAllRolesSet: ComputedRef<boolean>;
  doesContainOneVillagerSidedRole: ComputedRef<boolean>;
  doesContainOneWerewolfSidedRole: ComputedRef<boolean>;
  areRolesMinimumPlayersReached: ComputedRef<boolean>;
  isTwoSistersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  isThreeBrothersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  areAdditionalCardsSetForThiefIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForActorIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForAdditionalCardsDependantRoles: ComputedRef<boolean>;
  arePlayerGroupsSetForPrejudicedManipulatorIfPresent: ComputedRef<boolean>;
  canCreateGame: ComputedRef<boolean>;
  gameCreationValidationErrors: ComputedRef<string[]>;
  isRolePresentAndMinimumPlayersReached: (role: RoleWithMinInGame) => boolean;
};

function useCreateGameDtoValidation(createGameDto: Ref<CreateGameDto>): UseCreateGameDtoValidation {
  const rolesStore = useRolesStore();
  const { roles } = storeToRefs(rolesStore);

  const { t } = useI18n();

  const isMinimumPlayersReached = computed<boolean>(() => createGameDto.value.players.length >= MIN_PLAYERS_IN_GAME);

  const areAllRolesSet = computed<boolean>(() => createGameDto.value.players.every(player => player.role.name !== undefined));

  const doesContainOneVillagerSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "villagers"));

  const doesContainOneWerewolfSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "werewolves"));

  const areRolesMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const rolesWithMinimumPlayers = roles.value.filter((role): role is RoleWithMinInGame => role.minInGame !== undefined);

    return rolesWithMinimumPlayers.every(role => isRolePresentAndMinimumPlayersReached(role));
  });

  const isTwoSistersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "two-sisters") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const isThreeBrothersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "three-brothers") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const areAdditionalCardsSetForThiefIfPresent = computed<boolean>(() => {
    const isThiefPresent = createGameDto.value.players.some(player => player.role.name === "thief");
    if (!isThiefPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "thief") === true;
  });

  const areAdditionalCardsSetForActorIfPresent = computed<boolean>(() => {
    const isActorPresent = createGameDto.value.players.some(player => player.role.name === "actor");
    if (!isActorPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "actor") === true;
  });

  const areAdditionalCardsSetForAdditionalCardsDependantRoles = computed<boolean>(() => areAdditionalCardsSetForActorIfPresent.value &&
    areAdditionalCardsSetForThiefIfPresent.value);

  const arePlayerGroupsSetForPrejudicedManipulatorIfPresent = computed<boolean>(() => {
    const isPrejudicedManipulatorPresent = createGameDto.value.players.some(player => player.role.name === "prejudiced-manipulator");
    if (!isPrejudicedManipulatorPresent) {
      return true;
    }
    const minimumGroups = 2;
    const playersWithGroup = createGameDto.value.players.filter(player => player.group !== undefined) as CreateGamePlayerWithGroupDto[];
    const groups = groupByField(playersWithGroup, player => player.group) as Record<string, CreateGamePlayerWithGroupDto[]>;
    const areAllGroupsMinimumPlayersReached = Object.values(groups).every(group => group.length >= MIN_PLAYERS_IN_GROUP);

    return Object.keys(groups).length >= minimumGroups && areAllGroupsMinimumPlayersReached;
  });

  const canCreateGame = computed<boolean>(() => isMinimumPlayersReached.value &&
    areAllRolesSet.value &&
    doesContainOneVillagerSidedRole.value &&
    doesContainOneWerewolfSidedRole.value &&
    areRolesMinimumPlayersReached.value &&
    areAdditionalCardsSetForThiefIfPresent.value &&
    areAdditionalCardsSetForActorIfPresent.value &&
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value);

  const gameCreationValidationErrors = computed<string[]>(() => {
    const errors: Record<string, boolean> = {
      [t("composables.useCreateGameDtoValidation.minimumOfPlayersNotReached")]: !isMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.notAllRolesAreSet")]: !areAllRolesSet.value,
      [t("composables.useCreateGameDtoValidation.noVillagerSidedRole")]: !doesContainOneVillagerSidedRole.value,
      [t("composables.useCreateGameDtoValidation.noWerewolfSidedRole")]: !doesContainOneWerewolfSidedRole.value,
      [t("composables.useCreateGameDtoValidation.twoSistersMinimumPlayersNotReached")]: !isTwoSistersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.threeBrothersMinimumPlayersNotReached")]: !isThreeBrothersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.thiefAdditionalCardsNotSet")]: !areAdditionalCardsSetForThiefIfPresent.value,
      [t("composables.useCreateGameDtoValidation.actorAdditionalCardsNotSet")]: !areAdditionalCardsSetForActorIfPresent.value,
      [t("composables.useCreateGameDtoValidation.prejudicedManipulatorGroupsNotSet")]: !arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value,
    };

    return Object.keys(errors).filter((key: keyof typeof errors) => errors[key]);
  });

  function isRolePresentAndMinimumPlayersReached(role: RoleWithMinInGame): boolean {
    const playersWithRole = createGameDto.value.players.filter(player => player.role.name === role.name);
    const playersWithRoleCount = playersWithRole.length;

    return playersWithRoleCount === 0 || playersWithRoleCount >= role.minInGame;
  }
  return {
    isMinimumPlayersReached,
    areAllRolesSet,
    doesContainOneVillagerSidedRole,
    doesContainOneWerewolfSidedRole,
    areRolesMinimumPlayersReached,
    isTwoSistersRolePresentAndMinimumPlayersReached,
    isThreeBrothersRolePresentAndMinimumPlayersReached,
    areAdditionalCardsSetForThiefIfPresent,
    areAdditionalCardsSetForActorIfPresent,
    areAdditionalCardsSetForAdditionalCardsDependantRoles,
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent,
    canCreateGame,
    gameCreationValidationErrors,
    isRolePresentAndMinimumPlayersReached,
  };
}

export { useCreateGameDtoValidation };