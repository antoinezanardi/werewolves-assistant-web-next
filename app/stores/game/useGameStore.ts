import type { AsyncDataRequestStatus } from "nuxt/app";
import { defineStore } from "pinia";
import { unique } from "radash";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameEventsStore } from "~/stores/game/game-event/useGameEventsStore";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const {
    getGame: fetchGameFromApi,
    cancelGame: cancelGameFromApi,
    makeGamePlay: makeGamePlayFromApi,
  } = useFetchGames();

  const game = ref<Game>(new Game());
  const gameOptions = computed<GameOptions>(() => game.value.options);

  const gamePlayerGroups = computed<string[]>(() => unique(game.value.players.map(player => player.group)).filter(Boolean) as string[]);

  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const cancelingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const makingGamePlayStatus = ref<AsyncDataRequestStatus>("idle");

  const { resetGameEventIndex } = useGameEventsStore();

  function resetGame(): void {
    game.value = new Game();
  }

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }

  async function cancelGame(): Promise<void> {
    cancelingGameStatus.value = "pending";
    const canceledGame = await cancelGameFromApi(game.value._id);
    if (!canceledGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = canceledGame;
    cancelingGameStatus.value = "success";
  }

  async function makeGamePlay(makeGamePlayDto: MakeGamePlayDto): Promise<void> {
    makingGamePlayStatus.value = "pending";
    const playedGame = await makeGamePlayFromApi(game.value._id, makeGamePlayDto);
    if (!playedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = playedGame;
    makingGamePlayStatus.value = "success";
  }

  async function skipGamePlay(): Promise<void> {
    return makeGamePlay({});
  }
  return {
    game,
    gameOptions,
    gamePlayerGroups,
    fetchingGameStatus,
    cancelingGameStatus,
    makingGamePlayStatus,
    resetGame,
    fetchAndSetGame,
    cancelGame,
    makeGamePlay,
    skipGamePlay,
  };
});

export { useGameStore };