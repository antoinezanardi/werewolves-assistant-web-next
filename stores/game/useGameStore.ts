import type { AsyncDataRequestStatus } from "#app/composables/asyncData";
import { defineStore } from "pinia";

import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const { getGame: fetchGameFromApi } = useFetchGames();

  const game = ref<Game>(new Game());
  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }
  return {
    game,
    fetchingGameStatus,
    fetchAndSetGame,
  };
});

export { useGameStore };