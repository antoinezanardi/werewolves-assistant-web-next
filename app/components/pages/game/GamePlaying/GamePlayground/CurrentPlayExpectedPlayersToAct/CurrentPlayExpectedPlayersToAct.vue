<template>
  <div
    id="current-play-expected-players-to-act"
    class="w-full"
  >
    <PlayersHorizontalList
      id="player-horizontal-list"
      :players="expectedPlayersToAct"
    />
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import PlayersHorizontalList from "~/components/shared/game/player/PlayersHorizontalList/PlayersHorizontalList.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { useGameStore } from "~/stores/game/useGameStore";

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const expectedPlayersToAct = computed<Player[]>(() => game.value.currentPlay?.source.players ?? []);
</script>