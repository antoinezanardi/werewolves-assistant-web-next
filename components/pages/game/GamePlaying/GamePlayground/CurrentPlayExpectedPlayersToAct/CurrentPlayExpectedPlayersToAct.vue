<template>
  <div
    id="current-play-expected-players-to-act"
    class="mx-auto overflow-x-auto text-center w-full whitespace-nowrap"
  >
    <GlowCapture>
      <div
        v-for="player in expectedPlayersToAct"
        :key="player._id"
        class="expected-player-to-act inline-block text-center w-44"
      >
        <GlowElement>
          <RoleImage
            class="glow:border-gray-400 mx-auto"
            definition="normal"
            :role-name="player.role.current"
            sizes="100"
          />
        </GlowElement>

        <div class="truncate w-full">
          {{ player.name }}
        </div>
      </div>
    </GlowCapture>
  </div>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { useGameStore } from "~/stores/game/useGameStore";

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const expectedPlayersToAct = computed<Player[]>(() => game.value.currentPlay?.source.players ?? []);
</script>