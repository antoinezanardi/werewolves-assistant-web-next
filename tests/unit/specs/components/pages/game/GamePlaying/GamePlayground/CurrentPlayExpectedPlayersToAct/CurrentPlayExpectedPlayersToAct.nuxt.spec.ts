import { createTestingPinia } from "@pinia/testing";
import type { mount } from "@vue/test-utils";

import CurrentPlayExpectedPlayersToAct from "~/components/pages/game/GamePlaying/GamePlayground/CurrentPlayExpectedPlayersToAct/CurrentPlayExpectedPlayersToAct.vue";
import type PlayersHorizontalList from "~/components/shared/game/player/PlayersHorizontalList/PlayersHorizontalList.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameStore } from "~/stores/game/useGameStore";
import { createFakeGamePlaySource } from "~/tests/unit/utils/factories/composables/api/game/game-play/game-play-source/game-play-source.factory";
import { createFakeGamePlayCupidCharms } from "~/tests/unit/utils/factories/composables/api/game/game-play/game-play.factory";
import { createFakeGame } from "~/tests/unit/utils/factories/composables/api/game/game.factory";
import { createFakeCupidAlivePlayer } from "~/tests/unit/utils/factories/composables/api/game/player/player-with-role.factory";
import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";

describe("Current Play Expected Players To Act Component", () => {
  let wrapper: ReturnType<typeof mount<typeof CurrentPlayExpectedPlayersToAct>>;
  const expectedPlayersToAct = [
    createFakeCupidAlivePlayer({ name: "Toto" }),
    createFakeCupidAlivePlayer({ name: "Juju" }),
    createFakeCupidAlivePlayer({ name: "Doudou" }),
  ];
  const defaultGame = createFakeGame({
    currentPlay: createFakeGamePlayCupidCharms({
      source: createFakeGamePlaySource({
        players: expectedPlayersToAct,
        name: "cupid",
      }),
    }),
  });
  const testingPinia = { initialState: { [StoreIds.GAME]: { game: defaultGame } } };

  async function mountCurrentPlayExpectedPlayersToActComponent(): Promise<ReturnType<typeof mount<typeof CurrentPlayExpectedPlayersToAct>>> {
    return mountSuspendedComponent(CurrentPlayExpectedPlayersToAct, { global: { plugins: [createTestingPinia(testingPinia)] } });
  }

  beforeEach(async() => {
    wrapper = await mountCurrentPlayExpectedPlayersToActComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should not render the expected players to act when there is no current play.", async() => {
    const gameStore = useGameStore();
    gameStore.game = createFakeGame({ currentPlay: null });
    await nextTick();
    const playersHorizontalList = wrapper.findComponent<typeof PlayersHorizontalList>("#player-horizontal-list");

    expect(playersHorizontalList.props("players")).toStrictEqual<Player[]>([]);
  });

  it("should not render the expected players to act when there is no source players in current play.", async() => {
    const gameStore = useGameStore();
    gameStore.game = createFakeGame({ currentPlay: createFakeGamePlayCupidCharms({ source: createFakeGamePlaySource({ players: undefined }) }) });
    await nextTick();
    const playersHorizontalList = wrapper.findComponent<typeof PlayersHorizontalList>("#player-horizontal-list");

    expect(playersHorizontalList.props("players")).toStrictEqual<Player[]>([]);
  });

  it("should render the expected players to act when there is a current play.", () => {
    const playersHorizontalList = wrapper.findComponent<typeof PlayersHorizontalList>("#player-horizontal-list");

    expect(playersHorizontalList.props("players")).toStrictEqual<Player[]>(expectedPlayersToAct);
  });
});