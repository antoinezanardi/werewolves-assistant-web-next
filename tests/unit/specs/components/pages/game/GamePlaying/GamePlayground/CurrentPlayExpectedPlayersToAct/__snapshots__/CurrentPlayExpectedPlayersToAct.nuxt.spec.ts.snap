// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Current Play Expected Players To Act Component > should match snapshot when rendered. 1`] = `
"<div id="current-play-expected-players-to-act" class="mx-auto overflow-x-auto text-center w-full whitespace-nowrap">
  <div class="glow-capture" style="position: relative; --glow-size: 400px;">
    <div class="expected-player-to-act inline-block text-center w-44">
      <div class="grid glow" style="--glow-top: 0px; --glow-left: 0px;">
        <div class="" style="grid-area: 1/1/1/1;">
          <role-image-stub rolename="cupid" sizes="100" definition="normal" class="glow:border-gray-400 mx-auto"></role-image-stub>
        </div>
        <div class="glow-mask" glow="true" style="--glow-color: #f50057; grid-area: 1/1/1/1; pointer-events: none; mask: radial-gradient(var(--glow-size) var(--glow-size) at calc(var(--glow-x, -99999px) - var(--glow-left, 0px))
        calc(var(--glow-y, -99999px) - var(--glow-top, 0px)), #f50057 1%, transparent 50%);">
          <role-image-stub rolename="cupid" sizes="100" definition="normal" class="glow:border-gray-400 mx-auto"></role-image-stub>
        </div>
      </div>
      <div class="truncate w-full">Toto</div>
    </div>
    <div class="expected-player-to-act inline-block text-center w-44">
      <div class="grid glow" style="--glow-top: 0px; --glow-left: 0px;">
        <div class="" style="grid-area: 1/1/1/1;">
          <role-image-stub rolename="cupid" sizes="100" definition="normal" class="glow:border-gray-400 mx-auto"></role-image-stub>
        </div>
        <div class="glow-mask" glow="true" style="--glow-color: #f50057; grid-area: 1/1/1/1; pointer-events: none; mask: radial-gradient(var(--glow-size) var(--glow-size) at calc(var(--glow-x, -99999px) - var(--glow-left, 0px))
        calc(var(--glow-y, -99999px) - var(--glow-top, 0px)), #f50057 1%, transparent 50%);">
          <role-image-stub rolename="cupid" sizes="100" definition="normal" class="glow:border-gray-400 mx-auto"></role-image-stub>
        </div>
      </div>
      <div class="truncate w-full">Juju</div>
    </div>
    <div class="expected-player-to-act inline-block text-center w-44">
      <div class="grid glow" style="--glow-top: 0px; --glow-left: 0px;">
        <div class="" style="grid-area: 1/1/1/1;">
          <role-image-stub rolename="cupid" sizes="100" definition="normal" class="glow:border-gray-400 mx-auto"></role-image-stub>
        </div>
        <div class="glow-mask" glow="true" style="--glow-color: #f50057; grid-area: 1/1/1/1; pointer-events: none; mask: radial-gradient(var(--glow-size) var(--glow-size) at calc(var(--glow-x, -99999px) - var(--glow-left, 0px))
        calc(var(--glow-y, -99999px) - var(--glow-top, 0px)), #f50057 1%, transparent 50%);">
          <role-image-stub rolename="cupid" sizes="100" definition="normal" class="glow:border-gray-400 mx-auto"></role-image-stub>
        </div>
      </div>
      <div class="truncate w-full">Doudou</div>
    </div>
  </div>
</div>"
`;
